articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,249 @@
+
+#### Same key vs. different key
+
+If a ROS package declares a dependency on the `rosdep` key `python-foo` that maps on e.g. Ubuntu Xenial to the Python 2 package `foo`.
+In order to work for ROS 2 it needs to depend on the Python 3 version of the package instead.
+
+Currently the mapping in `rosdep` only depends on the platform and doesn't consider which ROS distribution is being used.
+
+If the migration to ROS 2 requires the key to be changed to `python3-foo` that implies that any package which depends on Python packages does need a modification in the manifest.
+
+The same conceptional problems already exist in ROS 1.
+When being built on platforms which use Python 3 they already map the keys to Python 3 packages.
+If this problem would be address within `rosdep` to consider the Python version as part of the platform that would allow Python 2 as well as Python 3 packages to be resolved using the same `rosdep` key.
+This would also allow building / using / testing ROS 1 with Python 3.
+
+Another example is the mapping of `rosdep` key to the Qt 4 or Qt 5 version of a dependency.
+A package may try to avoid changing the rosdep key used for their dependencies but rosdep only considers the platform without the context of the ROS distribution.
+
+**Proposal:** Change request for `rosdep` to consider information beyond the platform to determine the mapping.
+
+**Decision:** With that change the `rosdep` rules should share a single namespace (this is currently already the case).
+
+### Package manifest
+
+In ROS 1 two versions of the package manifest are supported: format 1 specified in [REP 127](http://www.ros.org/reps/rep-0127.html) and format 2 specified in [REP 140](http://www.ros.org/reps/rep-0140.html).
+While ROS 2 only supports format 2 there is no technical limitation behind that design decision.
+It simply reduces the effort to distinguish the different semantics of some depend tags in various parts of the software.
+
+It might be the case that a to-be-specified format 3 will be created to add support for the desired features to describe [dependency groups]() as well as declaring an [ABI version]().
+
+#### ROS dependencies
+
+While a ROS 1 package might depend on `roscpp` and `message_generation` a ROS 2 package depends on `rclcpp` and `rosidl_message_generation` instead.
+
+#### Build type
+
+Almost all build types for ROS 1 and ROS 2 are disjoint except the build type `cmake` which can be processed by both build tools.
+Since it is not desired to change existing `catkin` / `ament_cmake` packages to use the build type `cmake` those are currently different for each ROS version.
+
+### CMake version
+
+ROS 2 requires CMake version 3.5+.
+It should be easily possible to update existing CMake code to be compatible with various versions of CMake.
+
+### Python version
+
+ROS 2 requires Python version 3.5+.
+It should be easily possible to update existing Python code to be compatible with Python 2.7 as well as Python 3.5+.
```

Comment:
Is this proposing backporting all ROS 2 packages to support both python 2.7. I have only limited and very recent experience doing this with code that started life in Python 3, but it's a hefty amount of legwork. 

Off the top of my head

- Strings vs unicodes vs bytes don't appear to have a satisfying mutual API which means there will be a scattering of conditionals or wrapper functions required to support this.
- A bunch of changes were just made to ROS 2 code dropping things that were required in python 2 and optional in python 3. That work would all need to be reverted and future code would need to be written against 2/3 mutual python.

- How much did the Python C API change between Py2 and Py3? How much work would be involved in making rclpy py2 compatible?

All of this is doable but it's a significant amount of effort when Python 2.7 has a planned EOL in 2020. Even if the date gets extended, it seems like effort would be better spent determining the path forward for ROS 1 Python rather than backporting ROS 2 Python.

I guess the core question is: What is the benefit of making ROS 2 Python 2 compatible and is it worth the effort when it expires after two years?

Reply:
> What is the benefit of making ROS 2 Python 2 compatible and is it worth the effort when it expires after two years?

ROS 2 is only supporting Python 3.5+ and won't support older versions. The sentence was referring to ROS 1 which currently only support Python 2.7 but could be updated to support both.

Reply:
Clarified in ffc1b26f1098a948f882a3118d7b8426b7202b14.

Reply:
Okay. 😌 phew. 👍